cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,207 @@
+      session_ = std::move(s);
+      session_id_ = std::move(id);
+    }
+    return session_;
+  }
+
+  const CallHeaders& GetCallHeaders() const override { return headers_; }
+};
+
+std::vector<std::pair<std::string, std::string>>
+ServerSessionMiddlewareFactory::ParseCookieString(const std::string_view& s) {
+  const std::string list_sep = "; ";
+  const std::string pair_sep = "=";
+  const auto pair_sep_len = pair_sep.length();
+
+  std::vector<std::pair<std::string, std::string>> result;
+
+  size_t cur = 0;
+  while (cur < s.length()) {
+    const size_t end = s.find(list_sep, cur);
+    size_t len;
+    if (end == std::string::npos) {
+      // No (further) list delimiters
+      len = std::string::npos;
+      cur = s.length();
+    } else {
+      len = end - cur;
+      cur = end;
+    }
+    const std::string_view tok = s.substr(cur, len);
+
+    const size_t val_pos = tok.find(pair_sep);
+    if (val_pos == std::string::npos) {
+      // The cookie header is somewhat malformed; ignore the key and continue parsing
+      continue;
+    }
+    result.emplace_back(tok.substr(0, val_pos),
+                        tok.substr(val_pos + pair_sep_len, std::string::npos));
+  }
+
+  return result;
+}
+
+Status ServerSessionMiddlewareFactory::StartCall(
+    const CallInfo&, const CallHeaders& incoming_headers,
+    std::shared_ptr<ServerMiddleware>* middleware) {
+  std::string session_id;
+
+  const std::pair<CallHeaders::const_iterator, CallHeaders::const_iterator>&
+      headers_it_pr = incoming_headers.equal_range("cookie");
+  for (auto itr = headers_it_pr.first; itr != headers_it_pr.second; ++itr) {
+    const std::string_view& cookie_header = itr->second;
+    const std::vector<std::pair<std::string, std::string>> cookies =
+        ParseCookieString(cookie_header);
+    for (const std::pair<std::string, std::string>& cookie : cookies) {
+      if (cookie.first == kSessionCookieName) {
+        if (cookie.second.empty())
+          return Status::Invalid("Empty ", kSessionCookieName, " cookie value.");
+        session_id = std::move(cookie.second);
+      }
+    }
+    if (!session_id.empty()) break;
+  }
+
+  if (session_id.empty()) {
+    // No cookie was found
+    *middleware = std::make_shared<ServerSessionMiddlewareImpl>(this, incoming_headers);
+  } else {
+    const std::shared_lock<std::shared_mutex> l(session_store_lock_);
+    if (auto it = session_store_.find(session_id); it == session_store_.end()) {
+      return Status::Invalid("Invalid or expired ", kSessionCookieName, " cookie.");
+    } else {
+      auto session = it->second;
+      *middleware = std::make_shared<ServerSessionMiddlewareImpl>(
+          this, incoming_headers, std::move(session), session_id);
+    }
+  }
+
+  return Status::OK();
+}
+
+/// \brief Get a new, empty session option map and its id key.
+std::pair<std::string, std::shared_ptr<FlightSqlSession>>
+ServerSessionMiddlewareFactory::CreateNewSession() {
+  auto new_id = id_generator_();
```

Comment:
We may want to explicitly document that the generator should be thread-safe and that it must avoid collisions.

Reply:
Agreed, will do.

Reply:
`/// \param[in] id_gen A thread-safe, collision-free generator for session id strings.`

Reply:
Regarding both this and the other thread (ha) about this, probably worth erroring on collision (don't bother with a retry strategy) and a single SetSO call can fail once (ugly, but the generator's fault) instead of corrupting existing server session state (worse).

Reply:
```
  if (session_store_.count(new_id)) {
    // Collision
    return {"", NULLPTR};
  }
```

Results in not actually modifying the `session_store_` state and returns an empty/null shared_ptr to the caller, which is now documented.